File: Go/storage.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
)

const dataFile = "tasks.json"

// Storage управляет хранением задач
type Storage struct {
	Tasks  []*Task `json:"tasks"`
	NextID int     `json:"next_id"`
}

// NewStorage создает новое хранилище
func NewStorage() *Storage {
	return &Storage{
		Tasks:  make([]*Task, 0),
		NextID: 1,
	}
}

// Load загружает задачи из файла
func (s *Storage) Load() error {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		if os.IsNotExist(err) {
			return nil // Файл не существует - это нормально для первого запуска
		}
		return fmt.Errorf("ошибка чтения файла: %w", err)
	}

	if len(data) == 0 {
		return nil
	}

	err = json.Unmarshal(data, s)
	if err != nil {
		return fmt.Errorf("ошибка парсинга JSON: %w", err)
	}

	return nil
}

// Save сохраняет задачи в файл
func (s *Storage) Save() error {
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("ошибка сериализации: %w", err)
	}

	err = os.WriteFile(dataFile, data, 0644)
	if err != nil {
		return fmt.Errorf("ошибка записи файла: %w", err)
	}

	return nil
}

// AddTask добавляет новую задачу
func (s *Storage) AddTask(title, description string) *Task {
	task := NewTask(s.NextID, title, description)
	s.Tasks = append(s.Tasks, task)
	s.NextID++
	return task
}

// GetTask получает задачу по ID
func (s *Storage) GetTask(id int) *Task {
	for _, task := range s.Tasks {
		if task.ID == id {
			return task
		}
	}
	return nil
}

// DeleteTask удаляет задачу по ID
func (s *Storage) DeleteTask(id int) bool {
	for i, task := range s.Tasks {
		if task.ID == id {
			s.Tasks = append(s.Tasks[:i], s.Tasks[i+1:]...)
			return true
		}
	}
	return false
}

// ListTasks возвращает все задачи
func (s *Storage) ListTasks() []*Task {
	return s.Tasks
}

// FilterTasksByStatus возвращает задачи с указанным статусом
func (s *Storage) FilterTasksByStatus(status string) []*Task {
	filtered := make([]*Task, 0)
	for _, task := range s.Tasks {
		if task.Status == status {
			filtered = append(filtered, task)
		}
	}
	return filtered
}

// SearchTasks ищет задачи по тексту в названии или описании
func (s *Storage) SearchTasks(query string) []*Task {
	results := make([]*Task, 0)
	queryLower := strings.ToLower(query)

	for _, task := range s.Tasks {
		titleLower := strings.ToLower(task.Title)
		descLower := strings.ToLower(task.Description)

		if strings.Contains(titleLower, queryLower) || strings.Contains(descLower, queryLower) {
			results = append(results, task)
		}
	}
	return results
}

// FilterTasksByTag возвращает задачи с указанным тегом
func (s *Storage) FilterTasksByTag(tag string) []*Task {
	filtered := make([]*Task, 0)
	tagLower := strings.TrimSpace(strings.ToLower(tag))

	for _, task := range s.Tasks {
		for _, t := range task.Tags {
			if t == tagLower {
				filtered = append(filtered, task)
				break
			}
		}
	}
	return filtered
}

// GetAllTags возвращает список всех уникальных тегов
func (s *Storage) GetAllTags() []string {
	tagsMap := make(map[string]bool)
	for _, task := range s.Tasks {
		for _, tag := range task.Tags {
			tagsMap[tag] = true
		}
	}

	tags := make([]string, 0, len(tagsMap))
	for tag := range tagsMap {
		tags = append(tags, tag)
	}
	sort.Strings(tags)
	return tags
}

// SortTasks сортирует и возвращает задачи по указанному критерию
func (s *Storage) SortTasks(sortBy string) []*Task {
	tasks := make([]*Task, len(s.Tasks))
	copy(tasks, s.Tasks)

	switch sortBy {
	case "id":
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].ID < tasks[j].ID
		})
	case "created":
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].CreatedAt.Before(tasks[j].CreatedAt)
		})
	case "updated":
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].UpdatedAt.After(tasks[j].UpdatedAt)
		})
	case "status":
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].Status < tasks[j].Status
		})
	case "priority":
		priorityOrder := map[string]int{"high": 1, "medium": 2, "low": 3}
		sort.Slice(tasks, func(i, j int) bool {
			return priorityOrder[tasks[i].Priority] < priorityOrder[tasks[j].Priority]
		})
	}

	return tasks
}

// ExportToCSV экспортирует задачи в CSV файл
func (s *Storage) ExportToCSV(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("ошибка создания файла: %w", err)
	}
	defer file.Close()

	// Записываем BOM для корректного отображения UTF-8 в Excel
	file.WriteString("\xEF\xBB\xBF")

	// Заголовки
	fmt.Fprintln(file, "ID,Название,Описание,Статус,Приоритет,Дедлайн,Теги,Создано,Обновлено")

	// Данные
	for _, task := range s.Tasks {
		deadlineStr := ""
		if task.Deadline != nil {
			deadlineStr = task.Deadline.Format("02.01.2006 15:04")
		}

		tagsStr := ""
		if len(task.Tags) > 0 {
			tagStrings := make([]string, len(task.Tags))
			for i, tag := range task.Tags {
				tagStrings[i] = "#" + tag
			}
			tagsStr = strings.Join(tagStrings, ", ")
		}

		// Экранируем значения для CSV
		fmt.Fprintf(file, "%d,\"%s\",\"%s\",%s,%s,\"%s\",\"%s\",\"%s\",\"%s\"\n",
			task.ID,
			escapeCSV(task.Title),
			escapeCSV(task.Description),
			task.Status,
			task.Priority,
			deadlineStr,
			tagsStr,
			task.CreatedAt.Format("02.01.2006 15:04"),
			task.UpdatedAt.Format("02.01.2006 15:04"),
		)
	}

	return nil
}

// ExportToMarkdown экспортирует задачи в Markdown файл
func (s *Storage) ExportToMarkdown(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("ошибка создания файла: %w", err)
	}
	defer file.Close()

	fmt.Fprintf(file, "# Список задач\n\n")

	if len(s.Tasks) == 0 {
		fmt.Fprintln(file, "*Задач нет*")
		return nil
	}

	// Группируем по статусу
	statuses := []struct {
		key  string
		name string
	}{
		{"todo", "📋 К выполнению"},
		{"in_progress", "⚙️ В процессе"},
		{"done", "✅ Выполнено"},
	}

	for _, status := range statuses {
		statusTasks := make([]*Task, 0)
		for _, task := range s.Tasks {
			if task.Status == status.key {
				statusTasks = append(statusTasks, task)
			}
		}

		if len(statusTasks) > 0 {
			fmt.Fprintf(file, "## %s\n\n", status.name)

			for _, task := range statusTasks {
				// Приоритет
				priorityEmoji := map[string]string{
					"low":    "🟢",
					"medium": "🟡",
					"high":   "🔴",
				}
				priority := priorityEmoji[task.Priority]
				if priority == "" {
					priority = "⚪"
				}

				fmt.Fprintf(file, "### %s %s\n\n", priority, task.Title)
				fmt.Fprintf(file, "**ID:** %d  \n", task.ID)
				fmt.Fprintf(file, "**Описание:** %s  \n", task.Description)
				fmt.Fprintf(file, "**Приоритет:** %s  \n", task.Priority)

				// Дедлайн
				if task.Deadline != nil {
					fmt.Fprintf(file, "**Дедлайн:** %s  \n", task.Deadline.Format("02.01.2006 15:04"))
				}

				// Теги
				if len(task.Tags) > 0 {
					tagStrings := make([]string, len(task.Tags))
					for i, tag := range task.Tags {
						tagStrings[i] = "`#" + tag + "`"
					}
					fmt.Fprintf(file, "**Теги:** %s  \n", strings.Join(tagStrings, ", "))
				}

				fmt.Fprintf(file, "**Создано:** %s  \n", task.CreatedAt.Format("02.01.2006 15:04"))
				fmt.Fprintf(file, "**Обновлено:** %s  \n", task.UpdatedAt.Format("02.01.2006 15:04"))
				fmt.Fprintf(file, "\n---\n\n")
			}
		}
	}

	return nil
}

// escapeCSV экранирует кавычки в строке для CSV
func escapeCSV(s string) string {
	return strings.ReplaceAll(s, "\"", "\"\"")
}

File: Python/storage.py
import json
import os
import csv
from typing import List, Optional
from task import Task


class Storage:
    """Класс для управления хранением задач"""

    def __init__(self, filename: str = "tasks.json"):
        self.filename = filename
        self.tasks: List[Task] = []
        self.next_id = 1

    def load(self):
        """Загружает задачи из файла"""
        if not os.path.exists(self.filename):
            return

        try:
            with open(self.filename, 'r', encoding='utf-8') as f:
                data = json.load(f)
                self.next_id = data.get("next_id", 1)
                tasks_data = data.get("tasks", [])
                self.tasks = [Task.from_dict(task_data) for task_data in tasks_data]
        except (json.JSONDecodeError, KeyError) as e:
            print(f"Ошибка загрузки данных: {e}")
            self.tasks = []
            self.next_id = 1

    def save(self):
        """Сохраняет задачи в файл"""
        try:
            data = {
                "next_id": self.next_id,
                "tasks": [task.to_dict() for task in self.tasks]
            }
            with open(self.filename, 'w', encoding='utf-8') as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"Ошибка сохранения данных: {e}")

    def add_task(self, title: str, description: str) -> Task:
        """Добавляет новую задачу"""
        task = Task(self.next_id, title, description)
        self.tasks.append(task)
        self.next_id += 1
        return task

    def get_task(self, task_id: int) -> Optional[Task]:
        """Получает задачу по ID"""
        for task in self.tasks:
            if task.id == task_id:
                return task
        return None

    def delete_task(self, task_id: int) -> bool:
        """Удаляет задачу по ID"""
        for i, task in enumerate(self.tasks):
            if task.id == task_id:
                self.tasks.pop(i)
                return True
        return False

    def list_tasks(self) -> List[Task]:
        """Возвращает все задачи"""
        return self.tasks

    def filter_tasks_by_status(self, status: str) -> List[Task]:
        """Возвращает задачи с указанным статусом"""
        return [task for task in self.tasks if task.status == status]

    def search_tasks(self, query: str) -> List[Task]:
        """Ищет задачи по тексту в названии или описании"""
        query_lower = query.lower()
        return [
            task for task in self.tasks
            if query_lower in task.title.lower() or query_lower in task.description.lower()
        ]

    def filter_tasks_by_tag(self, tag: str) -> List[Task]:
        """Возвращает задачи с указанным тегом"""
        tag_lower = tag.lower().strip()
        return [task for task in self.tasks if tag_lower in task.tags]

    def get_all_tags(self) -> List[str]:
        """Возвращает список всех уникальных тегов"""
        all_tags = set()
        for task in self.tasks:
            all_tags.update(task.tags)
        return sorted(list(all_tags))

    def sort_tasks(self, sort_by: str) -> List[Task]:
        """Сортирует и возвращает задачи по указанному критерию"""
        tasks = self.tasks.copy()

        if sort_by == "id":
            tasks.sort(key=lambda t: t.id)
        elif sort_by == "created":
            tasks.sort(key=lambda t: t.created_at)
        elif sort_by == "updated":
            tasks.sort(key=lambda t: t.updated_at, reverse=True)
        elif sort_by == "status":
            tasks.sort(key=lambda t: t.status)
        elif sort_by == "priority":
            priority_order = {"high": 1, "medium": 2, "low": 3}
            tasks.sort(key=lambda t: priority_order.get(t.priority, 2))

        return tasks

    def export_to_csv(self, filename: str = "tasks.csv") -> bool:
        """Экспортирует задачи в CSV файл"""
        try:
            with open(filename, 'w', encoding='utf-8', newline='') as f:
                writer = csv.writer(f)
                # Заголовки
                writer.writerow(['ID', 'Название', 'Описание', 'Статус', 'Приоритет',
                                'Дедлайн', 'Теги', 'Создано', 'Обновлено'])

                # Данные
                for task in self.tasks:
                    deadline_str = task.deadline.strftime('%d.%m.%Y %H:%M') if task.deadline else ''
                    tags_str = ', '.join(['#' + t for t in task.tags]) if task.tags else ''

                    writer.writerow([
                        task.id,
                        task.title,
                        task.description,
                        task.status,
                        task.priority,
                        deadline_str,
                        tags_str,
                        task.created_at.strftime('%d.%m.%Y %H:%M'),
                        task.updated_at.strftime('%d.%m.%Y %H:%M')
                    ])
            return True
        except Exception as e:
            print(f"Ошибка экспорта в CSV: {e}")
            return False

    def export_to_markdown(self, filename: str = "tasks.md") -> bool:
        """Экспортирует задачи в Markdown файл"""
        try:
            with open(filename, 'w', encoding='utf-8') as f:
                f.write("# Список задач\n\n")

                if not self.tasks:
                    f.write("*Задач нет*\n")
                    return True

                # Группируем по статусу
                statuses = {
                    "todo": "📋 К выполнению",
                    "in_progress": "⚙️ В процессе",
                    "done": "✅ Выполнено"
                }

                for status, status_name in statuses.items():
                    status_tasks = [t for t in self.tasks if t.status == status]
                    if status_tasks:
                        f.write(f"## {status_name}\n\n")
                        for task in status_tasks:
                            # Приоритет
                            priority_emoji = {"low": "🟢", "medium": "🟡", "high": "🔴"}
                            priority_str = priority_emoji.get(task.priority, "⚪")

                            f.write(f"### {priority_str} {task.title}\n\n")
                            f.write(f"**ID:** {task.id}  \n")
                            f.write(f"**Описание:** {task.description}  \n")
                            f.write(f"**Приоритет:** {task.priority}  \n")

                            # Дедлайн
                            if task.deadline:
                                deadline_str = task.deadline.strftime('%d.%m.%Y %H:%M')
                                f.write(f"**Дедлайн:** {deadline_str}  \n")

                            # Теги
                            if task.tags:
                                tags_str = ', '.join(['`#' + t + '`' for t in task.tags])
                                f.write(f"**Теги:** {tags_str}  \n")

                            f.write(f"**Создано:** {task.created_at.strftime('%d.%m.%Y %H:%M')}  \n")
                            f.write(f"**Обновлено:** {task.updated_at.strftime('%d.%m.%Y %H:%M')}  \n")
                            f.write("\n---\n\n")
            return True
        except Exception as e:
            print(f"Ошибка экспорта в Markdown: {e}")
            return False
